common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala

```diff
@@ -0,0 +1,146 @@
+          transid.failed(this, start, s"initializiation failed with $t")
+      }
+      .flatMap { result =>
+        if (result.ok) {
+          Future.successful(result.interval)
+        } else if (result.interval.duration >= timeout) {
+          Future.failed(
+            InitializationError(
+              result.interval,
+              ActivationResponse.applicationError(Messages.timedoutActivation(timeout, true))))
+        } else {
+          Future.failed(
+            InitializationError(
+              result.interval,
+              ActivationResponse.processInitResponseContent(result.response, logging)))
+        }
+      }
+  }
+
+  def run(parameters: JsObject, environment: JsObject, timeout: FiniteDuration)(
+    implicit transid: TransactionId): Future[(Interval, ActivationResponse)] = {
+    val actionName = environment.fields.get("action_name").map(_.convertTo[String]).getOrElse("")
+    val start =
+      transid.started(this, LoggingMarkers.INVOKER_ACTIVATION_RUN, s"sending arguments to $actionName at $id $ip")
+
+    val parameterWrapper = JsObject("value" -> parameters)
+    val body = JsObject(parameterWrapper.fields ++ environment.fields)
+    callContainer("/run", body, timeout, retry = false)
+      .andThen { // never fails
+        case Success(r: RunResult) =>
+          transid.finished(
+            this,
+            start.copy(start = r.interval.start),
+            s"running result: ${r.toBriefString}",
+            endTime = r.interval.end)
+        case Failure(t) =>
+          transid.failed(this, start, s"run failed with $t")
+      }
+      .map { result =>
+        val response = if (result.interval.duration >= timeout) {
+          ActivationResponse.applicationError(Messages.timedoutActivation(timeout, false))
+        } else {
+          ActivationResponse.processRunResponseContent(result.response, logging)
+        }
+
+        (result.interval, response)
+      }
+  }
+
+  /**
+   * Makes an HTTP request to the container.
+   *
+   * Note that `http.post` will not throw an exception, hence the generated Future cannot fail.
+   *
+   * @param path relative path to use in the http request
+   * @param body body to send
+   * @param timeout timeout of the request
+   * @param retry whether or not to retry the request
+   */
+  protected def callContainer(path: String,
+                              body: JsObject,
+                              timeout: FiniteDuration,
+                              retry: Boolean = false): Future[RunResult] = {
+    val started = Instant.now()
+    val http = httpConnection.getOrElse {
+      val conn = new HttpUtils(s"${ip.asString}:${ip.port}", timeout, 1.MB)
+      httpConnection = Some(conn)
+      conn
+    }
+    Future {
+      http.post(path, body, retry)
+    }.map { response =>
+      val finished = Instant.now()
+      RunResult(Interval(started, finished), response)
+    }
+  }
+
+  def destroyApi()(implicit transid: TransactionId): Future[Unit] = {
```

Comment:
Why `destroyApi` vs. just `destroy`?

Reply:
I wasn't really sure how best to call `ContainerApi.destroy()` from `DockerContainer.destroy()`, since DockerContainer is now defined `with ContainerApi`. 
Ideas?

Reply:
Yeah I guess that's why I'd prefer it folded into either Container or have HttpContainer as an extension: Implementations can delegate to `super.destroy()`, which makes the impl. more transparent maybe?